### PR TITLE
implement double tests for existing resources

### PR DESCRIPTION
**Pull request by aufziehvogel**
- [x] Datacenter
- [x] Firewall
- [x] Firewall Actions
- [x] Floating IP
- [x] Floating IP Actions
- [x] Image
- [x] Image Actions
- [x] ISO
- [x] Location
- [x] Network
- [x] Network Actions
- [x] Server Actions
- [x] Server Type
- [x] SSH key
- [x] **Volume**
- [x] **Volume Actions**

bold: no tests existed previously

### CODE REVIEW

**Comment by aufziehvogel** [comment]
Changed two legacy tests:

- in floating IPs there was a test creating a floating IP with only `type`. According to the docs one of `home_location` or `server` is required. We now check such requirements, thus the test failed. I added `home_location`.
- there was a test in `firewall` checking that `set_rules(rules: nil)` is invalid. I changed this behaviour, because the API docs specify that the expected action to delete all firewall rules is `set_rules(rules: [])`. Allowing `set_rules(rules: nil)` to also delete all rules thus seemed natural to me.

File: lib/hcloud/server.rb [hunk 1]
```diff
@@ -37,10 +37,17 @@ def create_image(description: nil, type: nil)
...
+      raise Hcloud::Error::InvalidInput, 'no image given' if image.blank?
+
       prepare_request('actions/rebuild', j: { image: image }) { |j| j[:root_password] }
     end
 
     def change_type(server_type:, upgrade_disk: nil)
+      raise Hcloud::Error::InvalidInput, 'no server_type given' if server_type.blank?
+      # According to API docs, upgrade_disk is a required field. The docs do not
+      # mention that either true or false is a default.
+      raise Hcloud::Error::InvalidInput, 'no upgrade_disk given' if upgrade_disk.nil?
```
Comment (Kjarrigan) [inline lib/hcloud/server.rb:49]: In the frontend upgrade_disk defaults to false, because it's the "safe" way. You can always upscale the disk later if you need to, but you can never(!) downscale so true should never be the default. I wonder if we want to "copy" the default = false from the frontend because it's a safe choice but since this is an API client it's probably reasonable that the user should decide himself what he want to do here.

tl;dr - I would either set the default to false or remove the comment because it's "intended" that there is no default.

File: lib/hcloud/volume.rb [hunk 2]
```diff
@@ -18,6 +18,8 @@ class Volume
     has_actions
 
     def attach(server:, automount:)
```
Comment (Kjarrigan) [inline lib/hcloud/volume.rb:20]: Although not documented the default for `automount` is `false`, so I would reflect that default here as well.

File: lib/hcloud/volume.rb [hunk 3]
```diff
@@ -26,6 +28,8 @@ def detach
     end
 
     def resize(size:)
+      raise Hcloud::Error::InvalidInput, 'no size given' if size.nil?
```
Comment (Kjarrigan) [inline lib/hcloud/volume.rb:31]: size has to be an int (and actually greater that the current size). So I would change that check a bit to something like
```
   raise Hcloud::Error::InvalidInput, 'invalid size given' unless new_size.to_i > size
```

File: lib/hcloud/volume_resource.rb [hunk 4]
```diff
@@ -12,6 +12,12 @@ def [](arg)
     end
 
     def create(size:, name:, automount: nil, format: nil, location: nil, server: nil, labels: {})
+      raise Hcloud::Error::InvalidInput, 'no name given' if name.blank?
+      raise Hcloud::Error::InvalidInput, 'no size given' if size.nil?
```
Comment (Kjarrigan) [inline lib/hcloud/volume_resource.rb:16]: same here. I would do a `size.to_i > 10` check

File: lib/hcloud/volume_resource.rb [hunk 5]
```diff
@@ -12,6 +12,12 @@ def [](arg)
     end
 
     def create(size:, name:, automount: nil, format: nil, location: nil, server: nil, labels: {})
+      raise Hcloud::Error::InvalidInput, 'no name given' if name.blank?
```
Comment (Kjarrigan) [inline lib/hcloud/volume_resource.rb:15]: In network_resource you check for `raise Hcloud::Error::InvalidInput, 'no name given' if name.blank?` which is the "better check I think because it also checks for empty strings. So for "names" all over the place I would consistently use the blank? check (or always use the nil?)

**Review by Kjarrigan — COMMENTED** [review]
Part 1

File: spec/doubles/images.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+RSpec.shared_context 'images doubles' do
+  def new_image(kwargs = {})
+    {
+      id: Faker::Number.number,
+      name: Faker::Internet.slug,
+      description: Faker::Lorem.sentence,
+      # TODO: Or must disk_size be an integer?
+      disk_size: Faker::Number.positive,
+      image_size: Faker::Number.positive,
```
Comment (Kjarrigan) [inline spec/doubles/images.rb:11]: `disk_size` is always an integer
`image_size` is a float

File: spec/doubles/volumes.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+RSpec.shared_context 'volumes doubles' do
+  def new_volume(kwargs = {})
+    {
+      id: Faker::Number.number,
+      name: Faker::Internet.slug(glue: '-'),
+      created: Faker::Time.backward,
+      format: random_choice('ext4', 'xfs'),
```
Comment (Kjarrigan) [inline spec/doubles/volumes.rb:9]: `nil` is also a possible choice and actually the default

File: spec/hcloud/firewall_spec.rb [hunk 8]
```diff
@@ -1,212 +1,102 @@
...
+  context '#create' do
+    it 'handle missing name' do
+      expect { client.firewalls.create(name: nil) }.to(
+        raise_error(Hcloud::Error::InvalidInput)
+      )
+    end
+
+    context 'works' do
+      it 'with only name' do
+        params = { name: 'moo', labels: { 'key' => 'value' } }
```
Comment (Kjarrigan) [inline spec/hcloud/firewall_spec.rb:39]: the test-name is a bit misleading if I look at the params, because you provide name AND labels. Is this intended?

File: spec/integration/datacenter_spec.rb [hunk 9]
```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe 'Datacenter' do
```
Comment (Kjarrigan) [inline spec/integration/datacenter_spec.rb:5]: These tests look like the default `find_by_id_and_name` specs. Can't we use that default things too?

File: spec/integration/iso_spec.rb [hunk 10]
```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe 'ISO' do
```
Comment (Kjarrigan) [inline spec/integration/iso_spec.rb:5]: Same as with DC. This looks like it could be handled by the default `find_by_id_and_name` specs

File: spec/integration/location_spec.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe 'Location' do
```
Comment (Kjarrigan) [inline spec/integration/location_spec.rb:5]: Same as for DC

File: spec/integration/server_type_spec.rb [hunk 12]
```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe 'ServerType' do
```
Comment (Kjarrigan) [inline spec/integration/server_type_spec.rb:5]: No. 4 that seems to be redundant

File: spec/integration/ssh_key_spec.rb [hunk 13]
```diff
@@ -0,0 +1,128 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+REAL_KEY = 'ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAILh8GH'\
+  'JkJRgf3wuuUUQYG3UfqtVK56+FEXAOFaNZ659C m@x.com'
+
+describe 'SSHKey' do
```
Comment (Kjarrigan) [inline spec/integration/ssh_key_spec.rb:8]: Also has "special" find specs. Are these strictly necessary?

**Review by Kjarrigan — CHANGES_REQUESTED** [review]
Part 2
